fluster/cluster.py

```diff
@@ -38,7 +42,50 @@ def __init__(self,
                                       multiplier=penalty_box_wait_multiplier)
         self.active_clients = self._prep_clients(clients)
         self.initial_clients = {c.pool_id: c for c in clients}
+        self.clients = cycle(self.initial_clients.values())
         self._sort_clients()
+        # maintenance work is done each "tick" when iterating
+        self._tick_interval = 2 * len(clients)
+        self._ticks = 0
+
+    def __iter__(self):
+        """Updates active clients each time it's iterated through."""
+        self._prune_penalty_box()
+        return self
+
+    def __next__(self):
+        """Always returns a client, or raises an Exception if none are available."""
+        # raise Exception if no clients are available
+        if len(self.active_clients) == 0:
+            raise ClusterEmptyError('All clients are down.')
+
+        # always return something
+        nxt = None
+        while nxt is None:
+            nxt = self._next_helper()
+
+        self._ticks += 1
+        if self._ticks == self._tick_interval:
+            self._tick()
```

Comment:
Naming around this is sort of confusing. `self._ticks` and `self._tick`.  At first glance, I'd think that `_tick` incremented the counter.  Might be better to call it something like `self._cleanup` and `self._cleanup_interval`. Then you know each "tick" is a time when a client is gotten and returned.

All that said, I'm not sure ticks are necessary here. Since the penalty box is already time-based, shouldn't we be able to call it all the time with no real penalty?  It won't go and do the expensive operation (checking if down hosts are back) until time is up, so there's no harm in calling it and having it be a no-op.

Reply:
So that assumes the `heapq` is fairly efficient? If so, then yeah, that would simplify things and make it more resilient (no weird in-between stage to get stuck in)